### PR TITLE
add DefinitionRevison and make Componet/TraitDefinition revisionable

### PR DESCRIPTION
**Pull request by yangsoon**
fix https://github.com/oam-dev/kubevela/issues/1064

- [x] add DefinitionRevison types
- [x] fix the logic of  Componet/TraitDefinition controller
- [x] add test

### CODE REVIEW

File: legacy/charts/vela-core-legacy/crds/core.oam.dev_traitdefinitions.yaml [hunk 1]
```diff
@@ -7,13 +7,6 @@ metadata:
     controller-gen.kubebuilder.io/version: v0.2.4
   name: traitdefinitions.core.oam.dev
 spec:
-  additionalPrinterColumns:
-  - JSONPath: .spec.appliesToWorkloads
-    name: APPLIES-TO
-    type: string
-  - JSONPath: .metadata.annotations.definition\.oam\.dev/description
-    name: DESCRIPTION
-    type: string
```
Comment (wonderflow) [inline legacy/charts/vela-core-legacy/crds/core.oam.dev_traitdefinitions.yaml:16]: why deleted

File: legacy/charts/vela-core-legacy/crds/core.oam.dev_traitdefinitions.yaml [hunk 2]
```diff
@@ -25,8 +18,6 @@ spec:
     - trait
     singular: traitdefinition
   scope: Namespaced
-  subresources:
-    status: {}
   validation:
     openAPIV3Schema:
       description: A TraitDefinition registers a kind of Kubernetes custom resource as a valid OAM trait kind by referencing its CustomResourceDefinition. The CRD is used to validate the schema of the trait when it is embedded in an OAM ApplicationConfiguration.
```
Comment (wonderflow) [inline legacy/charts/vela-core-legacy/crds/core.oam.dev_traitdefinitions.yaml:23]: why

File: legacy/charts/vela-core-legacy/crds/core.oam.dev_traitdefinitions.yaml (hunk 1 above)
Comment (yangsoon) [reply legacy/charts/vela-core-legacy/crds/core.oam.dev_traitdefinitions.yaml:16]: sorry, it seem some wrong when generate crd file, i will fix it

**Review by wonderflow — APPROVED** [review]
Good job, let's add a doc for it

**Pull request merged**